### PR TITLE
json: limit float/double value which can be stored as integer

### PR DESCRIPTION
**Pull request by linev**
Before only condition value==floor(value) was checked.
But it can be true also for very large floats or doubles.
Up to certain limit it does not make sense.
Moreover, it can produce wrong output while value large 1e200 cannot be
stored at all

### CODE REVIEW

File: io/io/src/TBufferJSON.cxx [hunk 1]
```diff
@@ -3232,7 +3233,7 @@ void TBufferJSON::CompactFloatString(char* sbuf, unsigned len)
 void TBufferJSON::JsonWriteBasic(Float_t value)
 {
    char buf[200];
-   if (value == floor(value)) {
+   if ((value == floor(value)) && (TMath::Abs(value) < 1e15)) {
```
Comment (pcanal) [inline io/io/src/TBufferJSON.cxx]: Consider
```
std::abs(std::nearbyint(value) - value) > 0.00001f)
```

Comment (linev) [reply io/io/src/TBufferJSON.cxx]: I made new commit, but main condition remains:

    value == std::nearbyint(value)

Comment (pcanal) [reply io/io/src/TBufferJSON.cxx]: >  but main condition remains:

Isn't it wrong?  Testing floating point for equality is numerically unstable, isn't it?

**Comment by linev** [comment]
> Isn't it wrong? Testing floating point for equality is numerically unstable, isn't it?

Normal float representation is 1.2345e67
But in trivial cases (like 1 or 2 or 35) integer is much shorter.

Yes, `value == std::nearbyint(value)` not always return true for real integer, but this should not be a problem. In such case one just gets value like "5.0000001e1". And such value absolutely ok - other side should not expect integer at this place anyway. This is just to save space in produced JSON string.

**Comment by pcanal** [comment]
Fair enough (then my first recommendation is entirely useless), but please do add a comment expressing what is the intent (i.e. replace only *exact* match to integer or something like you say in your reply :) ). [Otherwise the code will 'eventually' be 'fixed' (wrongly) later on :) ] 

Thanks.

**Comment by linev** [comment]
I add comments

**Pull request merged**